### PR TITLE
port lint workflows from CircleCI to GHA

### PR DESCRIPTION
**Pull request by pmeier**
Per title. I don't expect any problems here so I'm ok with removing the lint workflows on CircleCI right away. LMK if you disagree so I can revert d31f3bbb77457cb664277736792557d203df53d0.

cc @seemethere

### CODE REVIEW

**Comment by NicolasHug** [comment]
Can you make these jobs fail so we can get a sense of the workflow for checking the errors?

**Comment by pmeier** [comment]
With c0cb6374e63c6fc98c4fae6b107f5e5cab5345f9:

- Python and configs: https://github.com/pytorch/vision/actions/runs/4375903581/jobs/7657254310
- C: https://github.com/pytorch/vision/actions/runs/4375903581/jobs/7657253653
- mypy: https://github.com/pytorch/vision/actions/runs/4375903581/jobs/7657255081

**Review by NicolasHug — APPROVED** [review]
Thanks Philip, looks like the detail of the  failure results are just one click away which is nice. LGTM!

**Comment by pmeier** [comment]
I looked into why the C lint job emits quite few of these warnings: https://github.com/pytorch/vision/actions/runs/4375903581/jobs/7657253653#step:10:196

> ./clang-format: /lib64/libtinfo.so.5: no version information available (required by ./clang-format)

According to [this SO answer](https://stackoverflow.com/a/72464818), `ncurses` from conda defaults channel simply doesn't provide this information. The apparent solution is to install from `conda-forge` instead. 

Since we are not building or testing here, this shouldn't be an issue and I tried that in c0006d0e33c7a2637ac1edc172c3448f1791301d. However the `clang-format` we are using depends on `libtinfo.so.5` that is provided by `ncurses=5`, which is over 6 years old at this point and also doesn't contain the version info. So this is patched in `libtinfo.so.6` from `conda-forge`, but that doesn't help us here.

This was not an issue on CircleCI since we were able to install it as system package on Ubuntu:

https://github.com/pytorch/vision/blob/e59cf64bb6eb4c3a50e0a76d8019fa4c4d5f2a15/.circleci/config.yml#L295-L297

The new runners use CentOS 7 and I was unable to install it through the system package manager.

Thus, since this is only a warning and doesn't affect the job otherwise, I'm ok to keep it as is. Any concerns @NicolasHug? @osalpekar I don't think there is an option to specify a docker image for the generic Linux job, is there?

**Comment by NicolasHug** [comment]
still OK for me. We rely on the clang job once in a blue moon anyway

File: .github/workflows/lint.yml [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+      - release/*
+  workflow_dispatch:
+
+jobs:
+  python-source-and-configs:
+    uses: pytorch/test-infra/.github/workflows/linux_job.yml@main
+    with:
+      repository: pytorch/vision
+      script: |
+        set -euo pipefail
```
Comment (osalpekar) [inline .github/workflows/lint.yml:18]: nit: -euxo pipefail

**Review by osalpekar — APPROVED** [review]
LGTM in general. As a longer term change, would be great to move linting to lintrunner like we have in PT Core (see more here: https://github.com/pytorch/pytorch/wiki/lintrunner). But this change is good-to-go!

**Review by seemethere — COMMENTED** [review]
LGTM, this looks great!

**Review by seemethere — APPROVED** [review]

File: .github/workflows/lint.yml (hunk 1 above)
Comment (pmeier) [reply .github/workflows/lint.yml:18]: See https://github.com/pytorch/vision/pull/7396#discussion_r1130028612.

**Pull request merged**